### ISSUE TITLE
When using Spring Boot 2.1.0 with azure-active-directory-spring-boot-starter get error on login page

### ISSUE THREAD

**Issue by dmusicant-eagle** [Microsoft/azure-spring-boot#457]
## Environment
- Spring boot starter:
  
  - active directory spring boot starter
  
- OS Type: MacOS
- Java version: 
  - 1.9 AND 1.8 both show the issue  

## Summary

When using Spring Boot 2.1.0 and `azure-active-directory-spring-boot-starter` 2.0.5 or 2.0.7, going to any URL (which kicks the user to "/login") results in the error:

```
java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.List (java.lang.String and java.util.List are in module java.base of loader 'bootstrap')
	at com.nimbusds.oauth2.sdk.util.URLUtils.serializeParameters(URLUtils.java:101) ~[oauth2-oidc-sdk-6.0.jar:6.0]
	at com.microsoft.aad.adal4j.AdalTokenRequest.toOAuthRequest(AdalTokenRequest.java:151) ~[adal4j-1.3.0.jar:1.3.0]
	...
```

## Reproduce steps

Set your dependencies as follows:

```
com.microsoft.azure:azure-active-directory-spring-boot-starter:2.0.7
org.springframework.boot:spring-boot-starter-web:2.1.0.RELEASE
org.springframework.boot:spring-boot-starter-thymeleaf:2.1.0.RELEASE
org.springframework.boot:spring-boot-starter-security:2.1.0.RELEASE
org.springframework.security:spring-security-oauth2-client:5.1.1.RELEASE
org.springframework.security:spring-security-oauth2-jose:5.1.1.RELEASE
org.thymeleaf.extras:thymeleaf-extras-springsecurity4:3.0.4.RELEASE
```
Run this sample: https://github.com/Microsoft/azure-spring-boot/tree/master/azure-spring-boot-samples/azure-active-directory-spring-boot-backend-sample

## Expected Results

That it would work without error.

## Actual Results

Throws the error:

```
java.lang.ClassCastException: class java.lang.String cannot be cast to class java.util.List (java.lang.String and java.util.List are in module java.base of loader 'bootstrap')
	at com.nimbusds.oauth2.sdk.util.URLUtils.serializeParameters(URLUtils.java:101) ~[oauth2-oidc-sdk-6.0.jar:6.0]
	at com.microsoft.aad.adal4j.AdalTokenRequest.toOAuthRequest(AdalTokenRequest.java:151) ~[adal4j-1.3.0.jar:1.3.0]
	at com.microsoft.aad.adal4j.AdalTokenRequest.executeOAuthRequestAndProcessResponse(AdalTokenRequest.java:84) ~[adal4j-1.3.0.jar:1.3.0]
	at com.microsoft.aad.adal4j.AuthenticationContext.acquireTokenCommon(AuthenticationContext.java:880) ~[adal4j-1.3.0.jar:1.3.0]
	at com.microsoft.aad.adal4j.AuthenticationContext.access$100(AuthenticationContext.java:68) ~[adal4j-1.3.0.jar:1.3.0]
	at com.microsoft.aad.adal4j.AuthenticationContext$1.call(AuthenticationContext.java:176) ~[adal4j-1.3.0.jar:1.3.0]
	at com.microsoft.aad.adal4j.AuthenticationContext$1.call(AuthenticationContext.java:165) ~[adal4j-1.3.0.jar:1.3.0]
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:834) ~[na:na]

```

**Comment by sergiogragera** [comment]
Same problem.

Do you know when is it possible to use the springboot 2.1.0 version? It is essential to be able to refresh the token automatically and it seems that the new version of oauth does it.

**Comment by ganeshchippada** [comment]
@sophiaso  i did not change any versions for the dependency but still getting the same error. By default it is pulling 5.1.1.RELEASE for the spring security. When i tried to change the spring-security-oauth2-client to 5.0.4.RELEASE, it started throwing errors due to version conflicts i believe. 

```


java.lang.IllegalStateException: Error processing condition on org.springframework.boot.autoconfigure.security.oauth2.client.servlet.OAuth2WebSecurityConfiguration.authorizedClientRepository
	at org.springframework.boot.autoconfigure.condition.SpringBootCondition.matches(SpringBootCondition.java:64) ~[spring-boot-autoconfigure-2.1.0.RELEASE.jar:2.1.0.RELEASE]
	at org.springframework.context.annotation.ConditionEvaluator.shouldSkip(ConditionEvaluator.java:108) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForBeanMethod(ConfigurationClassBeanDefinitionReader.java:181) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:141) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:117) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:327) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:232) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:275) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:95) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:691) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:528) ~[spring-context-5.1.2.RELEASE.jar:5.1.2.RELEASE]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:140) ~[spring-boot-2.1.0.RELEASE.jar:2.1.0.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:775) [spring-boot-2.1.0.RELEASE.jar:2.1.0.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) [spring-boot-2.1.0.RELEASE.jar:2.1.0.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:316) [spring-boot-2.1.0.RELEASE.jar:2.1.0.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1260) [spring-boot-2.1.0.RELEASE.jar:2.1.0.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1248) [spring-boot-2.1.0.RELEASE.jar:2.1.0.RELEASE]

```

**Comment by sophiaso** [comment]
@dmusicant-eagle, @sergiogragera, @ganeshchippada thanks for reporting issue, we have opened an [issue](https://github.com/Microsoft/azure-spring-boot/issues/466) tracking this, please expect a new release for Spring Boot 2.1.0.RELEASE in one week or two. Any help is also welcomed.

**Comment by jcainelli** [comment]
Good afternoon,

Does it work in any version of Spring 2?

**Comment by felipecrestani** [comment]
Hi, this work in any version?

**Comment by mkdirJava** [comment]
Still getting issues:
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>com.example</groupId>
	<artifactId>WilsonActiveDirectory</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<packaging>jar</packaging>

	<name>WilsonActiveDirectory</name>
	<description>Demo project for Spring Boot</description>

	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.0.3.RELEASE</version>
		<relativePath /> 
	</parent>

	<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
		<azure.version>2.0.5</azure.version>
		<java.version>1.8</java.version>
	</properties>

	<dependencies>



		<dependency>
			<groupId>org.springframework.security</groupId>
			<artifactId>spring-security-oauth2-client</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.security</groupId>
			<artifactId>spring-security-oauth2-jose</artifactId>
		</dependency>


		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-security</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>com.microsoft.azure</groupId>
			<artifactId>azure-active-directory-spring-boot-starter</artifactId>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-devtools</artifactId>
			<scope>runtime</scope>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>org.springframework.security</groupId>
			<artifactId>spring-security-test</artifactId>
			<scope>test</scope>
		</dependency>
	</dependencies>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>com.microsoft.azure</groupId>
				<artifactId>azure-spring-boot-bom</artifactId>
				<version>${azure.version}</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>


</project>


The response:

Whitelabel Error Page
This application has no explicit mapping for /error, so you are seeing this as a fallback.

Tue Dec 04 10:58:52 GMT 2018
There was an unexpected error (type=Forbidden, status=403).
Forbidden

**Comment by Incarnation-p-lee** [comment]
@mkdirJava
Thanks for your information, this should be compatibility issue for spring boot 2.1.x. We are planing the branch 2.0.x for spring boot 2.0.x and branch 2.1.x for spring boot 2.1.x. This is still under working, and will info you when ready.

**Comment by Incarnation-p-lee** [comment]
@mkdirJava 
For quickly validation, you can specific dependency in you POM like below to override the version 6.0 from spring security. Just one workaround here, need sdk side upgrade dependency.
```
<dependency>
    <groupId>com.nimbusds</groupId>
    <artifactId>oauth2-oidc-sdk</artifactId>
    <version>5.64.4</version>
</dependency>
```

**Comment by mkdirJava** [comment]
Will do this today, and let you know before end of the busines day

**Comment by mkdirJava** [comment]
Still the same issue:

here is the pom.xml


<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:schemaLocation="http://maven.apache.org/POM/4.0.0
http://maven.apache.org/xsd/maven-4.0.0.xsd">
<modelVersion>4.0.0</modelVersion>

<groupId>com.example</groupId>
<artifactId>WilsonActiveDirectory</artifactId>
<version>0.0.1-SNAPSHOT</version>
<packaging>jar</packaging>

<name>WilsonActiveDirectory</name>
<description>Demo project for Spring Boot</description>

<parent>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-parent</artifactId>
<version>2.0.3.RELEASE</version>
<relativePath /> 
</parent>

<properties>
<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
<azure.version>2.0.5</azure.version>
<java.version>1.8</java.version>
</properties>

<dependencies>

<dependency>
    <groupId>com.nimbusds</groupId>
    <artifactId>oauth2-oidc-sdk</artifactId>
    <version>5.64.4</version>
</dependency>


<dependency>
<groupId>org.springframework.security</groupId>
<artifactId>spring-security-oauth2-client</artifactId>
</dependency>
<dependency>
<groupId>org.springframework.security</groupId>
<artifactId>spring-security-oauth2-jose</artifactId>
</dependency>


<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-security</artifactId>
</dependency>
<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-web</artifactId>
</dependency>
<dependency>
<groupId>com.microsoft.azure</groupId>
<artifactId>azure-active-directory-spring-boot-starter</artifactId>
</dependency>

<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-devtools</artifactId>
<scope>runtime</scope>
</dependency>
<dependency>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-starter-test</artifactId>
<scope>test</scope>
</dependency>
<dependency>
<groupId>org.springframework.security</groupId>
<artifactId>spring-security-test</artifactId>
<scope>test</scope>
</dependency>
</dependencies>

<dependencyManagement>
<dependencies>
<dependency>
<groupId>com.microsoft.azure</groupId>
<artifactId>azure-spring-boot-bom</artifactId>
<version>${azure.version}</version>
<type>pom</type>
<scope>import</scope>
</dependency>
</dependencies>
</dependencyManagement>

<build>
<plugins>
<plugin>
<groupId>org.springframework.boot</groupId>
<artifactId>spring-boot-maven-plugin</artifactId>
</plugin>
</plugins>
</build>


</project>

**Comment by Incarnation-p-lee** [comment]
@mkdirJava 
Looks your pom leverage spring boot `2.0.3.RELEASE` which is different to this issue. If you meet some problem with 2.0.3.RELEASE, could you please open another issue for this ? Sorry for inconvenient.

**Comment by mkdirJava** [comment]
I am not understanding the last comment but I have tried again and am getting a different response when I go for 2.1.1.RELEASE

<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>com.example</groupId>
	<artifactId>WilsonActiveDirectory</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<packaging>jar</packaging>

	<name>WilsonActiveDirectory</name>
	<description>Demo project for Spring Boot</description>

	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<version>2.1.1.RELEASE</version>
		<relativePath /> 
	</parent>

	<properties>
		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
		<project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
		<azure.version>2.0.5</azure.version>
		<java.version>1.8</java.version>
	</properties>

	<dependencies>
	
	
		<dependency>
		    <groupId>com.nimbusds</groupId>
		    <artifactId>oauth2-oidc-sdk</artifactId>
		    <version>5.64.4</version>
		</dependency>


		<dependency>
			<groupId>org.springframework.security</groupId>
			<artifactId>spring-security-oauth2-client</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.security</groupId>
			<artifactId>spring-security-oauth2-jose</artifactId>
		</dependency>


		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-security</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-web</artifactId>
		</dependency>
		<dependency>
			<groupId>com.microsoft.azure</groupId>
			<artifactId>azure-active-directory-spring-boot-starter</artifactId>
		</dependency>

		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-devtools</artifactId>
			<scope>runtime</scope>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>org.springframework.security</groupId>
			<artifactId>spring-security-test</artifactId>
			<scope>test</scope>
		</dependency>
	</dependencies>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>com.microsoft.azure</groupId>
				<artifactId>azure-spring-boot-bom</artifactId>
				<version>${azure.version}</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>


</project>


on the response:


Login with OAuth 2.0
[authorization_request_not_found]




on the spring boot cmd

java.lang.NoSuchMethodError: org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser.<init>(Ljava/util/Set;Lorg/springframework/security/oauth2/core/oidc/OidcIdToken;Ljava/lang/String;)V
        at com.microsoft.azure.spring.autoconfigure.aad.AADOAuth2UserService.loadUser(AADOAuth2UserService.java:76) ~[azure-spring-boot-2.0.5.jar:na]
        at com.microsoft.azure.spring.autoconfigure.aad.AADOAuth2UserService.loadUser(AADOAuth2UserService.java:28) ~[azure-spring-boot-2.0.5.jar:na]
        at org.springframework.security.oauth2.client.oidc.authentication.OidcAuthorizationCodeAuthenticationProvider.authenticate(OidcAuthorizationCodeAuthenticationProvider.java:158) ~[spring-security-oauth2-client-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:174) ~[spring-security-core-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.oauth2.client.web.OAuth2LoginAuthenticationFilter.attemptAuthentication(OAuth2LoginAuthenticationFilter.java:186) ~[spring-security-oauth2-client-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:212) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestRedirectFilter.doFilterInternal(OAuth2AuthorizationRequestRedirectFilter.java:160) ~[spring-security-oauth2-client-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:116) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:100) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:74) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:56) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178) ~[spring-security-web-5.1.2.RELEASE.jar:5.1.2.RELEASE]
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:357) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:270) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:92) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107) ~[spring-web-5.1.3.RELEASE.jar:5.1.3.RELEASE]
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:199) ~[tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:490) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:791) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1417) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_172]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_172]
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-9.0.13.jar:9.0.13]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_172]



Given it mentions:

java.lang.NoSuchMethodError:


I am going to work on this and exclude

<dependency>
    <groupId>com.nimbusds</groupId>
    <artifactId>oauth2-oidc-sdk</artifactId>
    <version>5.64.4</version>
</dependency>


If anyone has a direction or update I would be gratful

**Comment by mkdirJava** [comment]
@Pan Li;
I understand why you overridden the Utils with :
<dependency>
    <groupId>com.nimbusds</groupId>
    <artifactId>oauth2-oidc-sdk</artifactId>
    <version>5.64.4</version>
</dependency>

I have set the version of spring boot to 2.0.3.RELEASE as the following mentions to do:

 sophiaso commented Nov 21, 2018

@jcainelli @felipecrestani, now it works with Spring Boot 2.0.3.RELEASE, please refer to this sample


I will continue to work on this getting the response of:
Redirecting you too many times

cmd mentions
java.lang.NoSuchMethodError: org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser.<init>(Ljava/util/Set;Lorg/springframework/security/oauth2/core/oidc/OidcIdToken;Ljava/lang/String;)V

**Comment by Incarnation-p-lee** [comment]
@mkdirJava 
Yes, you got the key point of dependency hell. For too many request redirect mostly should be the defect of web security configuration, you may trigger one auth and then redirect to some url, and the url trigger the auth again. Hope this helps.

For spring boot 2.1.x support, we are blocked by sdk side, nothing we can do for now excepting the workaround.

**Comment by mkdirJava** [comment]
@Incarnation-p-lee 
I would like to 

1. Highlight / ask something related to oauth2-oidc-sdk-6.0.jar 
2. Ask what is the work around here?

Point 1, I did some de-compiling and of the oauth2-oidc-sdk-6.0.jar ( found in the build )  specifically:
com.nimbusds.oauth2.sdk.util.URLUtils
looking the line number error on the cmd it mentions 101 being an issue and I can see it is a casting exception, rightly, said by the cmd. However I then look to the bitbucket site where I believe the code is coming from:

https://bitbucket.org/connect2id/oauth-2.0-sdk-with-openid-connect-extensions/src/2828a2e86f41b444294a1a59c23502f056333558/src/main/java/com/nimbusds/oauth2/sdk/util/URLUtils.java?at=release-6.0&fileviewer=file-view-default

I look up the class:
com.nimbusds.oauth2.sdk.util.URLUtils

and see that it has a different implementation. I am a little confused at the situation.

Point 2,
What is the temporary workaround is it to bump down spring version to 2.0.3.RELEASE? If so I have done so in the previous
I see a link in another comment I shall venture down this route whilst anyone can get back to me.

**Comment by Incarnation-p-lee** [comment]
@mkdirJava 
For details about the version conflict for API change, you can refer this issue https://github.com/AzureAD/azure-activedirectory-library-for-java/issues/248 from adal4j sdk.

**Comment by mkdirJava** [comment]
@Incarnation-p-lee 
Cheers Pan, will take a look

**Comment by dev4mobile** [comment]
please fix quickly

**Comment by Incarnation-p-lee** [comment]
@dmusicant-eagle @dev4mobile 
As our [azure spring boot 2.1.0](https://github.com/Microsoft/azure-spring-boot/releases) for spring boot 2.1.x has released last week, Could you please help to have a try for validating this issue ? If resolved could you please help to close it then ?
Thanks in advance!

**Comment by frjtrifork** [comment]
> @dmusicant-eagle @dev4mobile
> As our [azure spring boot 2.1.0](https://github.com/Microsoft/azure-spring-boot/releases) for spring boot 2.1.x has released last week, Could you please help to have a try for validating this issue ? If resolved could you please help to close it then ?
> Thanks in advance!

I have tested it today - using Spring Boot 2.1.1, and Azure Spring Boot 2.1.0

I still get the classcast exception if I do not explicitly override oauth2-oidc-sdk to 5.64.4.

`java.lang.ClassCastException: java.lang.String cannot be cast to java.util.List
	at com.nimbusds.oauth2.sdk.util.URLUtils.serializeParameters(URLUtils.java:101)
	at com.microsoft.aad.adal4j.AdalTokenRequest.toOAuthRequest(AdalTokenRequest.java:159)
	at com.microsoft.aad.adal4j.AdalTokenRequest.executeOAuthRequestAndProcessResponse(AdalTokenRequest.java:85)
	at com.microsoft.aad.adal4j.AuthenticationContext.acquireTokenCommon(AuthenticationContext.java:928)
	at com.microsoft.aad.adal4j.AcquireTokenCallable.execute(AcquireTokenCallable.java:70)
	at com.microsoft.aad.adal4j.AcquireTokenCallable.execute(AcquireTokenCallable.java:38)
	at com.microsoft.aad.adal4j.AdalCallable.call(AdalCallable.java:47)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)`

The problem arises in  (I have adal4j 1.6.0 on classpath - due to using 2.1.0 azure spring boot - I also tried setting adal4j 1.6.3 explicitly but that did not change anything)
com.microsoft.aad.adal4j.AdalTokenRequest#toOAuthRequest

As mentioned previously in this issue, Adal4j is built against nimbusds oauth2-oidc-sdk  5.64.4 - but spring boot security 2.1.1 brings in 6.0. 
Adal4j does not work with oauth2-oidc-sdk-6.0 (or any of the other 6.x releases)  

The problematic lines in adal4j are
`        final Map<String, String> params = this.grant.toParameters();
        httpRequest.setQuery(URLUtils.serializeParameters(params));
`
So the 'params' map is declared to have String values, but com.nimbusds.oauth2.sdk.util.URLUtils#serializeParameters in the oauth2-oidc-sdk-6.0.jar is declared as public static String serializeParameters(final Map<String,List<String>> params) 

Once nimbus tries iterating over the params map the classcastexception is thrown.

I did manage to get the 2.1.0 azure spring boot starter working with spring boot 2.1.1 - but had to explicitly set oauth2-oidc-sdk to version 5.64.4.

I use gradle - 
`
    compile("org.springframework.boot:spring-boot-starter-security:2.1.1.RELEASE")
    compile("org.springframework.security:spring-security-oauth2-client")
    compile("org.springframework.security:spring-security-oauth2-jose")
    compile("com.microsoft.azure:azure-active-directory-spring-boot-starter:2.1.0")
    compile("com.microsoft.azure:adal4j:1.6.3") // azure-active-directory-spring-boot-starter:2.1.0 brings in 1.6.0 - testing to see if 1.6.3 makes a difference
    compile("com.nimbusds:oauth2-oidc-sdk:5.64.4") // Adal4j 1.6.x requires 5.64 - spring boot / security depends on 6.x by default
`
Using that combination of dependencies I am able to validate tokens against azure directory service.

**Comment by Incarnation-p-lee** [comment]
@frjtrifork
OK. I see, we did not in involve that for gradle yet. As maven has something like dependency Management pom for all version control. I am not sure if gradle has similar feature. For quickly fix/workaround you can just override the `compile("com.nimbusds:oauth2-oidc-sdk:5.64.4")` with explicit version.

**Comment by Incarnation-p-lee** [comment]
Will close this issue for now, please feel free to reopen this if meet similar behavior.

**Comment by rbarrimond** [comment]
I've meet similar behavior! The workaround hasn't worked for me. Had to use JDBC as a workaround.

**Comment by Incarnation-p-lee** [comment]
@rbarrimond 
Could you please share more information about your issue ? like version, spring boot version, as well as stacktrace  for exceptions.

**Comment by rbarrimond** [comment]
> @rbarrimond
> Could you please share more information about your issue ? like version, spring boot version, as well as stacktrace for exceptions.

I can't anymore. Maven upgraded several jars and that resolved the issue based on updates to Spring and the Starter. I think what was happening is the jars were out of sync.
